spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/MappingSpannerReadConverter.java

```diff
@@ -73,163 +78,201 @@
-			Class sourceType,
-			String colName, PersistentPropertyAccessor accessor) {
-		Class targetType = spannerPersistentProperty.getType();
-		if (sourceType == null || !canConvert(sourceType, targetType)) {
-			return false;
-		}
-		BiFunction readFunction = singleItemReadMethodMapping
-				.get(ConversionUtils.boxIfNeeded(sourceType));
-		if (readFunction == null) {
-			return false;
+	private <R> boolean shouldSkipProperty(StructAccessor structAccessor,
+			SpannerPersistentProperty spannerPersistentProperty, Set<String> includeColumns,
+			boolean readAllColumns,
+			PreferredConstructor<? extends R, SpannerPersistentProperty> persistenceConstructor,
+			boolean allowMissingColumns) {
+		String columnName = spannerPersistentProperty.getColumnName();
+		boolean notRequiredByPartialRead = !readAllColumns && !includeColumns.contains(columnName);
+		boolean isConstructorParameter = persistenceConstructor != null
+				&& persistenceConstructor.isConstructorParameter(spannerPersistentProperty);
+
+		return notRequiredByPartialRead
+				|| isValidMissingColumn(structAccessor, allowMissingColumns, spannerPersistentProperty.getColumnName())
+				|| structAccessor.isNull(columnName)
+				|| isConstructorParameter;
+	}
+
+	private boolean isValidMissingColumn(StructAccessor structAccessor, boolean allowMissingColumns,
+			String columnName) {
+		if (!structAccessor.hasColumn(columnName)) {
+			if (!allowMissingColumns) {
+				throw new SpannerDataException(
+						"Unable to read column from Spanner results: "
+								+ columnName);
+			}
+			else {
+				return true;
+			}
 		}
-		accessor.setProperty(spannerPersistentProperty,
-				convert(readFunction.apply(struct, colName), targetType));
-		return true;
+		return false;
+	}
+
+	private <R> PersistentEntityParameterValueProvider<SpannerPersistentProperty> getParameterValueProvider(
+			SpannerPersistentEntity<? extends R> persistentEntity, StructAccessor structAccessor) {
+		return new PersistentEntityParameterValueProvider<>(
+				persistentEntity, new PropertyValueProvider<SpannerPersistentProperty>() {
+					@Override
+					public <T> T getPropertyValue(SpannerPersistentProperty property) {
+						return (T) structAccessor.get(property);
+					}
+				}, null);
 	}
 
-	private boolean attemptReadIterableValue(
-			SpannerPersistentProperty spannerPersistentProperty, Struct struct,
-			String colName, PersistentPropertyAccessor accessor) {
+	class StructAccessor {
+
+		private Struct struct;
 
-		Class innerType = ConversionUtils.boxIfNeeded(spannerPersistentProperty.getColumnInnerType());
-		if (innerType == null) {
-			return false;
+		StructAccessor(Struct struct) {
+			this.struct = struct;
 		}
 
-		// due to checkstyle limit of 3 return statments per function, this variable is
-		// used.
-		boolean valueSet = false;
+		public Object get(SpannerPersistentProperty spannerPersistentProperty) {
+			return ConversionUtils.isIterableNonByteArrayType(spannerPersistentProperty.getType())
```

Comment:
`StructAccessor` currently has multiple responsibilities: Access and conversion.

Typically, conversion is part of the converter itself. Accessors retrieve the actual value and use a converter method (See [`MongoDbPropertyValueProvider`](https://github.com/spring-projects/spring-data-mongodb/blob/239304160b35e594acd584a75556bc971d68ad4b/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MappingMongoConverter.java#L1323) and [`MappingMongoConverter.read(…)`](https://github.com/spring-projects/spring-data-mongodb/blob/239304160b35e594acd584a75556bc971d68ad4b/spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/MappingMongoConverter.java#L1374))

This way, you can apply the same conversion for properties and constructor properties and keep responsibilities simple.